<a href="https://colab.research.google.com/github/danielxf007/Fundamentos-Programacion-Especializacion-Analitica-y-Ciencia-de-Datos/blob/main/Prueba_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Prueba #2**
En la siguiente pureba se evalua de manera EXCLUSIVA lo que se solicita, cualquier otra respuesta adicional a ella será valorada, más no evaluada. Entregue la prueba en un repositorio de github antes de el domingo 18 de Febrero a las 23:59:59.

## **Ejercicio 1** (40%)

Construya un DataFrame a partir del siguiente diccionario

`exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'], 'score': [12.5, 9, 16.5, 10, 9, 20, 14.5, 12, 8, 19], 'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1], 'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}`

con las siguientes etiquetas para los índices

`labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']`

1. Seleccione solo las columnas *name* y *score*.
  
2. Filtre las filas donde el número de intentos (attemps) sea mayor que 2.

3. Filtre las filas donde el *score* se encuentre entre 15 y 20 (incluidos)

4. Cambie el *score* en la fila 'd' a 11.5

5. Agregue una nueva fila con etiqueta 'k' al dataframe con un valor que usted desee para cada columna

6. Borre la fila 'k' y entregue nuevamente el DataFrame original.

7. Ordene el DataFrame por nombre en orden ascendente.

8. Reemplace los valores de la columna *qualify* que contiene los valores *yes* y *no* por los valores booleanos True y False respectivamente.

In [ ]:
import pandas as pd
exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
              'score': [12.5, 9, 16.5, 10, 9, 20, 14.5, 12, 8, 19],
              'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
              'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

#se construye el data frame con etiquetas desde la letra 'a' a la letra 'j' para el índice
df = pd.DataFrame(exam_data, index=labels)
print(df)

        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   10.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes


1. Seleccionar solo las columnas name y score.

In [ ]:
#se seleccionan y se muestran unicamente las columnas name (nombre) y score (puntaje)
s_columnas=df[['name', 'score']]
print(s_columnas)

        name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
d      James   10.0
e      Emily    9.0
f    Michael   20.0
g    Matthew   14.5
h      Laura   12.0
i      Kevin    8.0
j      Jonas   19.0


2. Filtrar las filas donde el número de intentos (attemps) sea mayor que 2.

In [ ]:
#con el atributo 'loc' se realiza el filtro de la variable attemps que corresponde al numero de intentos, donde este sea mayor a 2
df.loc[df['attempts'] > 2]

name  score  attempts qualify
b     Dima    9.0         3      no
d    James   10.0         3      no
f  Michael   20.0         3     yes

3. Filtre las filas donde el score se encuentre entre 15 y 20 (incluidos)

In [ ]:
#con el atributo 'loc' se realiza el filtro de la variable score que corresponde al puntaje, donde este se encuentre entre 15 y 20 (incluidos)
df.loc[(df.score <= 20) & (df.score >= 15)]

name  score  attempts qualify
c  Katherine   16.5         2     yes
f    Michael   20.0         3     yes
j      Jonas   19.0         1     yes

4. Cambiar el score en la fila 'd' a 11.5

In [ ]:
#con el método 'at' se accede al valor específico que queremos modificar, en este caso 10.0 que será cambiado por 11.5
df.at['d', 'score'] = 11.5
print(df)

        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   11.5         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes


5. Agregar una nueva fila con etiqueta 'k' al dataframe con un valor que usted desee para cada columna

In [ ]:
#con el atributo 'loc' agregamos la nueva fila especificando los valores que se registrarán en ella
df.loc['k'] = ['Laura', 19.5, 2, 'yes']
print(df)

        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   11.5         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
k      Laura   19.5         2     yes


6. Borrar la fila 'k' y entregar nuevamente el DataFrame original.

In [ ]:
#con el método 'drop' eliminaremos la fila 'k' que fue agregada anteriormente
df = df.drop('k')
print(df)

        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   11.5         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes


7. Ordenar el DataFrame por nombre en orden ascendente.

In [ ]:
#con el método 'sort_values()' y el parámetro 'ascending' se ordena el dataframe por la variable 'name' de forma ascendente (A-Z)
df_ascendente = df.sort_values(by='name', ascending=True)
print(df_ascendente)

        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
e      Emily    9.0         2      no
d      James   11.5         3      no
j      Jonas   19.0         1     yes
c  Katherine   16.5         2     yes
i      Kevin    8.0         2      no
h      Laura   12.0         1      no
g    Matthew   14.5         1     yes
f    Michael   20.0         3     yes


8. Reemplazar los valores de la columna qualify que contiene los valores yes y no por los valores booleanos True y False respectivamente.

In [ ]:
#con el método 'replace' se reemplazan los valores de la varibale 'qualify' por valores booleanos: 'yes' por True y 'no' por False
df['qualify'] = df['qualify'].replace({'yes': True, 'no': False})
print(df)

        name  score  attempts  qualify
a  Anastasia   12.5         1     True
b       Dima    9.0         3    False
c  Katherine   16.5         2     True
d      James   11.5         3    False
e      Emily    9.0         2    False
f    Michael   20.0         3     True
g    Matthew   14.5         1     True
h      Laura   12.0         1    False
i      Kevin    8.0         2    False
j      Jonas   19.0         1     True


## **Ejercicio 2** (30%)

Importe el dataset `athletes` del repositorio del curso y realice las siguientes tareas:

1. Importarlo como nombre athletes
2. Filtre el dataframe tal que solo se quede con los atletas que hayan ganado al menos una medalla.
3. Calcule la masa muscular por fila con una función como

$$I_{m.m.} = \frac{Weight}{Height^2}$$

4. Agrupe los datos por calculando la media por deporte para cada una del resto de características. ¿Cuál es el deporte en el que los jugadores tienen mayor peso promedio?

5. ¿Cuál es el pais con más participación femenina?

# Solución Ejercicio 2

###1. Se Importa el dataset con el nombre "athletes"

In [ ]:
import pandas as pd
# Se importa el dataset athletes.csv
athletes = pd.read_csv('https://raw.githubusercontent.com/tomasate/Datos_Clases/main/Datos_1/athletes.csv')
athletes

id                name nationality     sex       dob  height  \
0      736041664      A Jesus Garcia         ESP    male  10/17/69    1.72   
1      532037425          A Lam Shin         KOR  female   9/23/86    1.68   
2      435962603         Aaron Brown         CAN    male   5/27/92    1.98   
3      521041435          Aaron Cook         MDA    male    1/2/91    1.83   
4       33922579          Aaron Gate         NZL    male  11/26/90    1.81   
...          ...                 ...         ...     ...       ...     ...   
11533  265605954  Zurian Hechavarria         CUB  female   8/10/95    1.64   
11534  214461847      Zuzana Hejnova         CZE  female  12/19/86    1.73   
11535   88361042             di Xiao         CHN    male   5/14/91    1.85   
11536  900065925   le Quoc Toan Tran         VIE    male    4/5/89    1.60   
11537  711404576      le Roux Hamman         RSA    male    1/6/92    1.85   

       weight          sport  gold  silver  bronze  
0        64.0      athletics     0       0       0  
1        56.0        fencing     0       0       0  
2        79.0      athletics     0       0       1  
3        80.0      taekwondo     0       0       0  
4        71.0        cycling     0       0       0  
...       ...            ...   ...     ...     ...  
11533    58.0      athletics     0       0       0  
11534    63.0      athletics     0       0       0  
11535   100.0      wrestling     0       0       0  
11536    56.0  weightlifting     0       0       0  
11537    70.0      athletics     0       0       0  

[11538 rows x 11 columns]

###2. Se filtra el dataframe tal que solo se quede con los atletas que hayan ganado al menos una medalla (Bronce, Plata y Oro).

In [ ]:
# Se obtienen del DataFrame los atletas que obtuvieron al menos una medalla
athletes[athletes['gold'] + athletes['silver'] + athletes['bronze'] >= 1]

id                         name nationality     sex       dob  \
2      435962603                  Aaron Brown         CAN    male   5/27/92   
6      266237702                Aaron Russell         USA    male    6/4/93   
14     162792594               Abbey Weitzeil         USA  female   12/3/96   
48     962468808  Abdoulrazak Issoufou Alfaga         NIG    male  12/26/94   
55     969824503           Abdullah Alrashidi         IOA    male   8/21/63   
...          ...                          ...         ...     ...       ...   
11487  334126873                  Zhiyong Shi         CHN    male  10/10/93   
11491  105609876                        Zi He         CHN  female  12/10/90   
11503  593183472                  Zivko Gocic         SRB    male   8/22/82   
11509  836633838                      Zoe Lee         GBR  female  12/15/85   
11511  343671816                Zoe de Toledo         GBR  female   7/17/87   

       height  weight          sport  gold  silver  bronze  
2        1.98    79.0      athletics     0       0       1  
6        2.05    98.0     volleyball     0       0       1  
14       1.78    68.0       aquatics     1       1       0  
48       2.07    98.0      taekwondo     0       1       0  
55       1.83    84.0       shooting     0       0       1  
...       ...     ...            ...   ...     ...     ...  
11487    1.68    69.0  weightlifting     1       0       0  
11491    1.60    52.0       aquatics     0       1       0  
11503    1.93    93.0       aquatics     1       0       0  
11509    1.76    70.0         rowing     0       1       0  
11511    1.72    58.0         rowing     0       1       0  

[1857 rows x 11 columns]

###3. Se calcula la masa muscular por fila con una función como

$$I_{m.m.} = \frac{Weight}{Height^2}$$

In [ ]:
# Se crea una función ind_masa_musc() que calcula el índice de masa muscular

def ind_masa_musc(row):
  ind_mm = row.weight / (row.height ** 2)
  return ind_mm

athletes.apply(ind_masa_musc, axis = 1)

0        21.633315
1        19.841270
2        20.151005
3        23.888441
4        21.672110
           ...    
11533    21.564545
11534    21.049818
11535    29.218408
11536    21.875000
11537    20.452885
Length: 11538, dtype: float64

###4. Se agrupan los datos (calculando la media) por deporte para cada una del resto de características. ¿Cuál es el deporte en el que los jugadores tienen mayor peso promedio?

In [ ]:
# Se agrupan los datos (calculando la media) por deporte para cada una de las características
athletes.groupby('sport').mean().sort_values(by = 'weight',ascending=False)

<ipython-input-47-365f0a67356f>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  athletes.groupby('sport').mean().sort_values(by = 'weight',ascending=False)


id    height     weight      gold    silver  \
sport                                                                      
basketball         5.036080e+08  1.918715  87.750000  0.083333  0.083333   
handball           4.957886e+08  1.837080  83.710602  0.079890  0.082645   
volleyball         4.844270e+08  1.896527  80.102094  0.072917  0.070312   
weightlifting      5.004174e+08  1.666628  79.980620  0.058140  0.058140   
rowing             5.095929e+08  1.849477  79.938318  0.087751  0.087751   
rugby sevens       5.228671e+08  1.755623  78.723906  0.083333  0.080000   
wrestling          4.937765e+08  1.723000  77.742120  0.050992  0.050992   
canoe              5.019726e+08  1.781006  77.015291  0.081571  0.081571   
judo               4.903510e+08  1.735366  76.876316  0.035714  0.035714   
shooting           4.877244e+08  1.728668  73.905263  0.038462  0.038462   
tennis             4.971817e+08  1.802251  73.162304  0.040816  0.040816   
aquatics           4.997725e+08  1.788338  72.301644  0.083045  0.077509   
archery            5.047630e+08  1.735952  72.190476  0.062500  0.062500   
golf               4.741152e+08  1.743761  71.443478  0.016667  0.016667   
sailing            5.085194e+08  1.761478  71.169355  0.039474  0.039474   
fencing            5.527768e+08  1.772449  70.661224  0.085366  0.089431   
hockey             5.178987e+08  1.733981  68.900463  0.078704  0.074074   
badminton          4.901661e+08  1.745529  68.774390  0.046512  0.046512   
football           5.064191e+08  1.745341  68.433962  0.058920  0.058920   
taekwondo          5.245064e+08  1.797480  68.088000  0.062500  0.062500   
cycling            4.978055e+08  1.746667  67.820717  0.051429  0.053333   
athletics          5.016411e+08  1.751922  67.717727  0.027931  0.027084   
equestrian         4.618328e+08  1.748704  67.493023  0.067568  0.067568   
modern pentathlon  4.242136e+08  1.762361  65.958333  0.027778  0.027778   
table tennis       5.024409e+08  1.721845  65.177515  0.046512  0.046512   
triathlon          4.347202e+08  1.732110  60.633028  0.018182  0.018182   
gymnastics         5.121383e+08  1.637840  54.278997  0.092593  0.092593   
boxing             4.731657e+08  1.747745        NaN  0.045455  0.045455   

                     bronze  
sport                        
basketball         0.083333  
handball           0.079890  
volleyball         0.072917  
weightlifting      0.058140  
rowing             0.087751  
rugby sevens       0.083333  
wrestling          0.101983  
canoe              0.084592  
judo               0.071429  
shooting           0.038462  
tennis             0.040816  
aquatics           0.073356  
archery            0.062500  
golf               0.016667  
sailing            0.039474  
fencing            0.089431  
hockey             0.076389  
badminton          0.046512  
football           0.058920  
taekwondo          0.125000  
cycling            0.055238  
athletics          0.026238  
equestrian         0.067568  
modern pentathlon  0.027778  
table tennis       0.046512  
triathlon          0.018182  
gymnastics         0.092593  
boxing             0.090909

In [ ]:
# Se reduce la vista para mostrar unicamente los deportes y el peso promedio por deporte
athletes.groupby('sport').mean()[['weight']].sort_values(by = 'weight',ascending=False)

<ipython-input-48-f03462e1296d>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  athletes.groupby('sport').mean()[['weight']].sort_values(by = 'weight',ascending=False)


weight
sport                       
basketball         87.750000
handball           83.710602
volleyball         80.102094
weightlifting      79.980620
rowing             79.938318
rugby sevens       78.723906
wrestling          77.742120
canoe              77.015291
judo               76.876316
shooting           73.905263
tennis             73.162304
aquatics           72.301644
archery            72.190476
golf               71.443478
sailing            71.169355
fencing            70.661224
hockey             68.900463
badminton          68.774390
football           68.433962
taekwondo          68.088000
cycling            67.820717
athletics          67.717727
equestrian         67.493023
modern pentathlon  65.958333
table tennis       65.177515
triathlon          60.633028
gymnastics         54.278997
boxing                   NaN

El **deporte** en el que los jugadores tienen un **mayor peso promedio** es en el **basketball** con un peso medio de **87.75(um)**.

Seguido por balonmano con un preso medio aproximado de 83.71(um)

###5. ¿Cuál es el pais con más participación femenina?

In [ ]:
athletes[athletes['sex'] == 'female'].groupby(['nationality','sex']).size().sort_values(ascending = False)

nationality  sex   
USA          female    303
CHN          female    251
AUS          female    220
BRA          female    216
GER          female    204
                      ... 
SOM          female      1
SSD          female      1
STP          female      1
ASA          female      1
AFG          female      1
Length: 202, dtype: int64

El país con mayor participación femenina es **Estados Unidos (USA)** con total de **303 mujeres**.

Seguido por China (CHN) con una participación de 251 mujeres

## **Ejercicio 3 (30%)**

Importe el dataset de nba del repositorio de clase y responda (hint: Elimine las instancias con registros faltantes):

1. Cuáles son los equipos que pagan los salarios más altos (presente el resultado como un dataFrame con el nombre del equipo y la cantidad de dinero de la nómina)
2. Separe la variable *Age* en tres grupos y cree una nueva columna llamada *Age-Group* con esta información.<
3. Construya una tabla dinámica que muestre el salario promedio de los jugadores por equipo y posición para para los cinco equipos que pagan mejores salarios.

In [1]:
import pandas as pd

In [2]:
nba_df = pd.read_csv('https://raw.githubusercontent.com/tomasate/Datos_Clases/main/Datos_1/nba.csv')
nba_df

Unnamed: 0           Name            Team  Number Position   Age Height  \
0             0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   
1             1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   
2             2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   
3             3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   
4             4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   
..          ...            ...             ...     ...      ...   ...    ...   
453         453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   
454         454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   
455         455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   
456         456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   
457         457            NaN             NaN     NaN      NaN   NaN    NaN   

     Weight            College     Salary  
0     180.0              Texas  7730337.0  
1     235.0          Marquette  6796117.0  
2     205.0  Boston University        NaN  
3     185.0      Georgia State  1148640.0  
4     231.0                NaN  5000000.0  
..      ...                ...        ...  
453   203.0             Butler  2433333.0  
454   179.0                NaN   900000.0  
455   256.0                NaN  2900000.0  
456   231.0             Kansas   947276.0  
457     NaN                NaN        NaN  

[458 rows x 10 columns]

In [3]:
nba_df2 = nba_df.dropna()
nba_df2.isnull().values.any()

False

1.Cuáles son los equipos que pagan los salarios más altos (presente el resultado como un dataFrame con el nombre del equipo y la cantidad de dinero de la nómina)

In [4]:
nba_salaries = nba_df2.groupby(['Team']).agg({'Salary':'sum'}).sort_values(by=['Salary'], ascending=False).head()
nba_salaries

Salary
Team                             
Los Angeles Clippers   93906914.0
Oklahoma City Thunder  81515298.0
Golden State Warriors  80535517.0
Cleveland Cavaliers    74068189.0
Chicago Bulls          73265800.0

2.Separe la variable Age en tres grupos y cree una nueva columna llamada Age-Group con esta información

In [5]:
nba_df3 = nba_df2.copy()
nba_df3['Age-Group'] = pd.cut(nba_df2.Age, bins = 3, labels = ['group1', 'group2', 'group3'])
nba_df3

Unnamed: 0           Name            Team  Number Position   Age Height  \
0             0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   
1             1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   
3             3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   
6             6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   
7             7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   
..          ...            ...             ...     ...      ...   ...    ...   
449         449    Rodney Hood       Utah Jazz     5.0       SG  23.0    6-8   
451         451  Chris Johnson       Utah Jazz    23.0       SF  26.0    6-6   
452         452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   
453         453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   
456         456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   

     Weight        College     Salary Age-Group  
0     180.0          Texas  7730337.0    group1  
1     235.0      Marquette  6796117.0    group1  
3     185.0  Georgia State  1148640.0    group1  
6     235.0            LSU  1170960.0    group1  
7     238.0        Gonzaga  2165160.0    group1  
..      ...            ...        ...       ...  
449   206.0           Duke  1348440.0    group1  
451   206.0         Dayton   981348.0    group1  
452   234.0       Kentucky  2239800.0    group1  
453   203.0         Butler  2433333.0    group1  
456   231.0         Kansas   947276.0    group1  

[364 rows x 11 columns]

3.Construya una tabla dinámica que muestre el salario promedio de los jugadores por equipo y posición para para los cinco equipos que pagan mejores salarios.

In [6]:
pd.options.display.float_format = '{:.2f}'.format

In [7]:
nba_df3[nba_df3['Team'].isin(nba_salaries.index)].groupby(['Team', 'Position']).agg({'Salary':'mean'})

Salary
Team                  Position            
Chicago Bulls         C        13400000.00
                      PF        3578739.67
                      PG       11171532.00
                      SF        1958160.00
                      SG        5717549.25
Cleveland Cavaliers   C         7768435.00
                      PF       13941014.50
                      PG        6551592.33
                      SF         947276.00
                      SG        3349079.00
Golden State Warriors C         6541249.33
                      PF       14260870.00
                      PG       11370786.00
                      SF        3766367.40
                      SG        8224138.00
Los Angeles Clippers  C        10394801.00
                      PF        6655482.00
                      PG       12289745.50
                      SF        3662923.75
                      SG        4639893.33
Oklahoma City Thunder C         6303142.67
                      PF        2606520.00
                      PG        9382869.00
                      SF        8599620.67
                      SG        3207057.50